app/src/main/java/com/urbandroid/sleep/addon/stats/model/socialjetlag/SocialJetlagStats.java

```diff
@@ -64,17 +75,175 @@ public SocialJetlagStats narrow(Date from, Date to) {
+        TreeMap<String, BitSet> sleepStateByDateMap = new TreeMap<>();
+        for(int j = 0; j < awakeTimes.size(); ++j){
+            Interval awakeInterval = awakeTimes.get(j);
+
+            BitSet sleepState;
+
+            DateTime awakeIntervalStartTime = awakeInterval.getStart();
+            String awakeIntervalFrom = awakeIntervalStartTime.toString("yyyy-MM-dd");
+
+            if(sleepStateByDateMap.containsKey(awakeIntervalFrom)){
+                sleepState = sleepStateByDateMap.get(awakeIntervalFrom);
+            } else{
+                sleepState = new BitSet(1440); // 24 hrs * 60 minutes
+            }
+
+            int startIdx = awakeIntervalStartTime.getMinuteOfDay();
+            int endIdx = awakeInterval.getEnd().getMinuteOfDay();
+            sleepState.set(startIdx, endIdx);
+
+            sleepStateByDateMap.put(awakeIntervalFrom, sleepState);
+        }
+
+        return sleepStateByDateMap;
+    }
+
+    /**
+     * Calculates SRI score given 2 days
+     * It is the (total number of minutes in which a user is in a consistent sleep state) / 1440 mins
+     * Consistent state: prevDay[i] == nextDay[i]
+     * Inconsistent state: prevDay[i] =/= nextDay[i]
+     * @param prevDay
+     * @param nextDay
+     * @return
+     */
+    public float calculateSRI(BitSet prevDay, BitSet nextDay){
+        BitSet minsInconsistentSleepState = (BitSet) prevDay.clone();
+        minsInconsistentSleepState.andNot(nextDay); // store inconsistent sleep state mins in prevDayTmp
+
+        return (1.0f - (minsInconsistentSleepState.length()/1440.f)); // 1 - totalMinsOfInconsistentSleepState
+    }
+
+    /**
+     * Calculates average SRI across multiple, continuous dates
+     * @param sleepStateByDateMap
+     * @return
+     */
+    public float calculateAverageSRI(TreeMap<String, BitSet> sleepStateByDateMap){
+
+        float cumulativeSRI = 0;
+
+        Set keys = sleepStateByDateMap.keySet();
+
+        Iterator nextDayIt = keys.iterator();
+        // iterate through pairs of days to calculate the SRI
+        if (nextDayIt.hasNext()) {
+            nextDayIt.next();
+            for (Iterator prevDayIt = keys.iterator(); prevDayIt.hasNext() && nextDayIt.hasNext();) {
+
+                String prevDateKey = (String) prevDayIt.next();
+                String nextDateKey =  (String) nextDayIt.next();
+                DateTime prevDate = new DateTime(prevDateKey);
+                DateTime nextDate = new DateTime(nextDateKey);
+
+                if(isNextDate(prevDate, nextDate)){
+                    BitSet prevDaySleepStates = sleepStateByDateMap.get(prevDateKey);
+                    BitSet nextDaySleepStates = sleepStateByDateMap.get(nextDateKey);
+
+                    // extract the date and month and if it is one
+                    float sriScore = calculateSRI(prevDaySleepStates, nextDaySleepStates);
+                    cumulativeSRI += sriScore;
+
+                } else{ // TODO: throw an error?
```

Comment:
If the data is missing several days of data, would it make more since simply to continue calculating where the new sequence of contiguous days begins?

Reply:
good point - this is also a design choice - because SRI in theory requires a continuous set of days. If I skip an entire week you're not really comparing the day before and after; let's keep this for now and discuss what would be best